### PR TITLE
string -> expression -> string -> expression

### PR DESCRIPTION
**Pull request by clintropolis**
# Description

This PR adds a new method to the `Expr` interface, `Expr.stringify()`. which produces _parseable_ expression strings so that any `Expr` tree can be converted back into a `String` which can later be parsed into an equivalent expression.

Prior to this PR, not all `Expr` which could exist at evaluation time were actually parseable, specifically empty numeric arrays and arrays with null elements. To make all `Expr` able to satisfy the `stringify` contract, the grammar has been updated to support these constructs. Empty arrays may now be defined like so: `<STRING>[]`, `<DOUBLE>[]`, `<LONG>[]`, and arrays like `[null, 1, 2]` are now able to be correctly parsed from an expression string. The explicit typing syntax for arrays extends beyond empty arrays, so things like `<LONG>[1, 2, 3]` are also valid (and equivalent to `[1, 2, 3]`).

For testing, I modified the majority of expression unit tests assertions to also ensure that parsing the stringified form of a parsed expression produces the same results with and without flattening, so that nearly every expression that is tested also tests round tripping back to string.

Tagged with design review to discuss syntax for the empty arrays.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `Expr.java`
 * `Expr.g4`
 * `ExprListenerImpl.java`

### CODE REVIEW

File: core/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java [hunk 1]
```diff
@@ -188,9 +195,15 @@ public void exitLogicalAndOrExpr(ExprParser.LogicalAndOrExprContext ctx)
...
-    Long[] values = new Long[ctx.LONG().size()];
+    Long[] values = ctx.longArrayBody() == null
+                    ? new Long[0]
+                    : new Long[ctx.longArrayBody().longElement().size()];
     for (int i = 0; i < values.length; i++) {
-      values[i] = Long.parseLong(ctx.LONG(i).getText());
+      if (ctx.longArrayBody().longElement(i).getText().equalsIgnoreCase(Expr.NULL_LITERAL)) {
+        values[i] = null;
+      } else {
+        values[i] = Long.parseLong(ctx.longArrayBody().longElement(i).getText());
```
Comment (jihoonson) [inline core/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java]: Wondering if the array elements should be casted if they are not longs.

File: core/src/main/java/org/apache/druid/math/expr/ExprMacroTable.java [hunk 2]
```diff
@@ -132,17 +141,29 @@ private BindingDetails supplyAnalyzeInputs()
...
       analyzeInputsSupplier = Suppliers.memoize(this::supplyAnalyzeInputs);
     }
 
+    @Override
+    public String stringify()
+    {
+      return StringUtils.format(
+          "%s(%s)",
+          name,
+          Joiner.on(", ").join(args.stream().map(Expr::stringify).iterator())
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/ExprMacroTable.java]: changed

File: core/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java (hunk 1 above)
Comment (clintropolis) [reply core/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java]: Yeah, it seems more druidish to do that, I've modified explicitly typed numeric arrays to use a more permissive parsing with a couple of methods I've added to `Numbers` utility class, so `<LONG>[1, 1.2, 3]` and `<DOUBLE>[1, 2, 3]` will coerce the elements as `new Long[]{1L, 1L, 3L} ` and `new Double[]{1.0, 2.0, 3.0}` respectively, as one might hope. However, I am not converting string literals, so things like `<LONG>[1, 2, '3']` are still not valid at this time.

I also made explicit string arrays more permissive, and can accept any type of literal as an element and will convert them all to strings, `<STRING>['hello', 1, 1.2]` -> `new String[]{"hello", "1", "1.2"}`.

Comment (jihoonson) [reply core/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java]: Thanks, sounds good.

**Pull request merged**